src/callback_handler/crypto.rs

```diff
@@ -82,217 +137,165 @@ impl CertificatePool {
+        root_ca_cert_param.extended_key_usages = [ExtendedKeyUsagePurpose::CodeSigning].to_vec();
+        root_ca_cert_param
+    }
 
-    #[test]
-    fn certificate_is_trusted() {
-        // use the correct CA chain
+    fn generate_certificate(
+        is_ca: bool,
+        ca: Option<CertifiedKey<KeyPair>>,
+        not_before: Option<OffsetDateTime>,
+        not_after: Option<OffsetDateTime>,
+    ) -> Result<CertifiedKey<KeyPair>> {
+        let is_ca_param = if is_ca {
+            rcgen::IsCa::Ca(rcgen::BasicConstraints::Unconstrained)
+        } else {
+            rcgen::IsCa::NoCa
+        };
+        let cert_param = get_base_ca_cert_params(is_ca_param, not_before, not_after);
+        let signing_key = KeyPair::generate().expect("Failed to generate certificate key pair");
+
+        let cert = match (is_ca, ca) {
+            // root ca
+            (true, None) => cert_param.self_signed(&signing_key),
+            // intermediate ca or end entity certificate
+            (_, Some(certified_key)) => {
+                let issuer =
+                    Issuer::from_ca_cert_der(certified_key.cert.der(), certified_key.signing_key)
+                        .expect("Failed to generate issuer");
+                cert_param.signed_by(&signing_key, &issuer)
+            }
+            _ => {
+                unimplemented!();
+            }
+        }
+        .expect("Failed to create certificate");
+        Ok(CertifiedKey { cert, signing_key })
+    }
+
+    fn generate_certificate_chain(
+        not_before: Option<OffsetDateTime>,
+        not_after: Option<OffsetDateTime>,
+    ) -> (Certificate, Vec<Certificate>) {
+        let root_ca =
+            generate_certificate(true, None, None, None).expect("Failed to create root CA");
+        let root_ca_cert = Certificate {
+            encoding: CertificateEncoding::Pem,
+            data: root_ca.cert.pem().as_bytes().to_vec(),
+        };
+
+        let intermediate_ca = generate_certificate(true, Some(root_ca), None, None)
+            .expect("Failed to create intermediate CA");
         let ca_cert = Certificate {
             encoding: CertificateEncoding::Pem,
-            data: ROOT_CA1_PEM.as_bytes().to_vec(),
+            data: intermediate_ca.cert.pem().as_bytes().to_vec(),
         };
-        let cert_chain = vec![ca_cert];
-        let cert = Certificate {
+
+        let end_entity_certificate =
+            generate_certificate(false, Some(intermediate_ca), not_before, not_after)
+                .expect("Failed to create end entity cert");
+        let end_entity_cert = Certificate {
             encoding: CertificateEncoding::Pem,
-            data: INTERMEDIATE_CA1_PEM.as_bytes().to_vec(),
+            data: end_entity_certificate.cert.pem().as_bytes().to_vec(),
         };
+
+        // use the correct CA chain
+        let cert_chain = vec![ca_cert, root_ca_cert];
+        (end_entity_cert, cert_chain)
```

Comment:
With the adoption of the rustls-webpki to verify the certificate, it's not possible to have a CA certificate in the beginning of the chain. Therefore, this function generated a root ca, intermediate ca and a end entity (leaf) certiciate